Dockerfile

```
@@ -1,60 +1,35 @@
-# Two stage Dockerfile to build OpenWPM
-# Stage 1 builds the extension, stage 2 builds the main OpenWPM image
-
-FROM node:10 as extension
-WORKDIR /usr/src/app
-
-# The extension needs to run for example the TypeScript transpiler
-# to generate the JavaScript code of the extension. This must be done as root
-# as long as the directory of the extension is only writeable as root.
-RUN npm config set unsafe-perm true
-
-COPY automation/Extension/firefox/. ./
-COPY automation/Extension/webext-instrumentation/. ../webext-instrumentation
-RUN npm install
-RUN npm run build
-RUN cp dist/openwpm-1.0.zip openwpm.xpi
-
-# Stage 2, build the main OpenWPM image
 FROM krallin/ubuntu-tini:bionic
 
-WORKDIR /opt/OpenWPM
+SHELL ["/bin/bash", "-c"]
 
-# This is just a performance optimization and can be skipped by non-US
-# based users
+# Update ubuntu and setup conda
+# adapted from: https://hub.docker.com/r/conda/miniconda3/dockerfile
 RUN sed -i'' 's/archive\.ubuntu\.com/us\.archive\.ubuntu\.com/' /etc/apt/sources.list
-
-RUN apt-get clean -y && rm -r /var/lib/apt/lists/* -vf && apt-get clean -y && apt-get update -y && apt-get upgrade -y && apt-get install sudo -y
-
-# Install the Ubuntu packages as well as firefox and the geckodriver first
-COPY ./install-system.sh .
-RUN ./install-system.sh
+RUN apt-get clean -qq \
+    && rm -r /var/lib/apt/lists/* -vf \
+    && apt-get clean -qq \
+    && apt-get update -qq \
+    && apt-get upgrade -qq \
+    # git and make for `npm install`, wget for `install-miniconda`
+    && apt-get install wget git make -qq \
+    # deps to run firefox
+    && apt-get install libgtk-3-0 libx11-xcb1 libdbus-glib-1-2 libxt6 -qq
+
+ENV HOME /opt
+COPY scripts/install-miniconda.sh .
+RUN ./install-miniconda.sh
+ENV PATH $HOME/miniconda/bin:$PATH
+
+# Install OpenWPM
+WORKDIR /opt/OpenWPM
+COPY . .
+RUN ./install.sh
```

By having this run after COPY we need to rebuild the enviroment every time any file in the entire repo changes, which takes a significant amount of time and might break the workflow of people that develop in a docker image.
```suggestion
COPY enviroment.yaml scripts/conda_setup.sh .
RUN ./scripts/conda_setup.sh
COPY automation/extension .
RUN ./scripts/build-extension.sh
COPY . .
```
Where `scripts/conda-setup.sh` would be the first part of `install.sh`

---

I agree to a certain extent. In particular, every time I changed the Dockerfile I needed to rerun a lot. I considered adding more to .dockerignore, but wanted to check in on that.

My reasoning behind what I have now is as little code duplication as I could possibly squeeze so we’re only updating one thing when we do maintenance.

Here’s what it would look like - it would mean some duplication from install.sh - but it’s not that bad. It would look like this:

```dockerfile
COPY scripts .  # We could copy scripts one by one but it’s probably not worth it, they won’t change often.
RUN ./scripts/install-miniconda.sh
ENV PATH $HOME/miniconda/bin:$PATH
COPY environment.yaml .
RUN conda env create --force -q -f environment.yaml
ENV PATH $HOME/miniconda/envs/openwpm/bin:$PATH
COPY automation/extension .
RUN ./scripts/build-extension.sh
RUN ./scripts/install-firefox.sh
... move firefox
COPY automation crawler.py demo.py . # Or `COPY . .`
```

I wanted to clarify what you mean by `scripts/conda-setup.sh` - there’s only one line - I don’t think it needs to be broken out into a script.

> Might break the workflow of people that develop in a docker image

It’s not that much code and not that much duplication, so I’m okay about making the change, but who’s actually developing in a docker image? (Or who would be now). When I was developing the docker file, and didn’t want to do all the waiting, I updated the docker file myself to include the lines from `install.sh`. We could also include instructions about this in the README and keep the main docker file clean of duplicated code.

---

> I wanted to clarify what you mean by scripts/conda-setup.sh - there’s only one line - I don’t think it needs to be broken out into a script.

Yeah, it's just a oneliner I didn't think that one through properly, sorry.

> but who’s actually developing in a docker image?

I was for a longer period of time that I'd like to admit because the pinned dependencies weren't available for Python3.8, but since that's no longer an issue I'm mostly thinking about people that develop on Windows but still want to run OpenWPM.

[Pycharm for example allows configuring a docker container as a remote interpreter](https://www.jetbrains.com/help/pycharm/using-docker-as-a-remote-interpreter.html) and [VSCode allows something similiar](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers)
That's also why I appreciate now having the dev dependencies be part of the docker image.

---

Please hold for windows update....

---

I've successfully run through the dev pipeline using WSL1 and 2, suggest we recommend that and don't recommend / support developing with docker.  I will open an issue to document developing on windows.